File: packages/quran_bridge_ios/ios/Classes/AudioController.swift
import Foundation
import AVFoundation

protocol AudioControllerDelegate: AnyObject {
    func audioController(_ controller: AudioController, didEmit state: [String: Any])
}

/// Manages audio playback and emits state maps to the EventChannel delegate.
final class AudioController: NSObject {

    weak var delegate: AudioControllerDelegate?

    private var player: AVPlayer?
    private var timeObserver: Any?
    private var currentSurah: Int?
    private var currentAyah: Int?

    // MARK: - Commands

    func play(from map: [String: Any]) throws {
        guard let surah    = map["surahNumber"] as? Int,
              let reciter  = map["reciterId"]   as? String else {
            throw QuranBridgeError.invalidArgument("AudioRequest is malformed")
        }

        currentSurah = surah
        currentAyah  = map["fromAyah"] as? Int

        // TODO: integrate quran-ios library here — resolve audio URL from reciter + surah
        let urlString = audioURL(reciter: reciter, surah: surah)
        guard let url = URL(string: urlString) else {
            throw QuranBridgeError.audioError("Could not resolve audio URL for reciter '\(reciter)'")
        }

        emit(status: "loading")

        stop()

        let item = AVPlayerItem(url: url)
        player = AVPlayer(playerItem: item)

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerDidFinish),
            name: .AVPlayerItemDidPlayToEndTime,
            object: item
        )

        addProgressObserver()

        player?.play()
        emit(status: "playing")
    }

    func pause() {
        player?.pause()
        emit(status: "paused")
    }

    func stop() {
        removeProgressObserver()
        NotificationCenter.default.removeObserver(self)
        player?.pause()
        player = nil
        emit(status: "stopped")
    }

    // MARK: - Private

    private func addProgressObserver() {
        timeObserver = player?.addPeriodicTimeObserver(
            forInterval: CMTime(seconds: 0.5, preferredTimescale: 600),
            queue: .main
        ) { [weak self] time in
            guard let self else { return }
            let posMs = Int(time.seconds * 1000)
            let durMs = Int((self.player?.currentItem?.duration.seconds ?? 0) * 1000)
            self.emit(status: "playing", positionMs: posMs, durationMs: durMs)
        }
    }

    private func removeProgressObserver() {
        if let observer = timeObserver {
            player?.removeTimeObserver(observer)
            timeObserver = nil
        }
    }

    @objc private func playerDidFinish() {
        stop()
        emit(status: "stopped")
    }

    private func emit(
        status: String,
        positionMs: Int? = nil,
        durationMs: Int? = nil
    ) {
        var state: [String: Any] = ["status": status]
        if let s = currentSurah  { state["surahNumber"] = s }
        if let a = currentAyah   { state["ayahNumber"]  = a }
        if let p = positionMs    { state["positionMs"]  = p }
        if let d = durationMs    { state["durationMs"]  = d }
        delegate?.audioController(self, didEmit: state)
    }

    // TODO: integrate quran-ios library here — replace with real URL resolution
    private func audioURL(reciter: String, surah: Int) -> String {
        let paddedSurah = String(format: "%03d", surah)
        return "https://cdn.islamic.network/quran/audio-surah/128/\(reciter)/\(paddedSurah).mp3"
    }
}

// MARK: - Typed errors

enum QuranBridgeError: Error {
    case invalidArgument(String)
    case notFound(String)
    case audioError(String)

    var flutterError: FlutterError {
        switch self {
        case .invalidArgument(let msg):
            return FlutterError(code: "INVALID_ARGS",  message: msg, details: nil)
        case .notFound(let msg):
            return FlutterError(code: "NOT_FOUND",     message: msg, details: nil)
        case .audioError(let msg):
            return FlutterError(code: "AUDIO_ERROR",   message: msg, details: nil)
        }
    }
}

File: packages/quran_bridge_android/android/src/main/kotlin/dev/quran/bridge/AudioController.kt
package dev.quran.bridge

import android.content.Context
import android.media.MediaPlayer
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/// Manages audio playback and streams state via EventChannel.
class AudioController(private val context: Context) : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null
    private var player: MediaPlayer? = null
    private val handler = Handler(Looper.getMainLooper())
    private var currentSurah: Int? = null
    private var currentAyah: Int? = null

    private val progressRunnable = object : Runnable {
        override fun run() {
            player?.takeIf { it.isPlaying }?.let {
                emit("playing", positionMs = it.currentPosition, durationMs = it.duration)
                handler.postDelayed(this, 500)
            }
        }
    }

    // ── StreamHandler ─────────────────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ── Commands ──────────────────────────────────────────────────────────────

    @Suppress("UNCHECKED_CAST")
    fun play(map: Map<String, Any>) {
        val surah   = map["surahNumber"] as? Int
            ?: throw QuranBridgeException.invalidArgs("surahNumber")
        val reciter = map["reciterId"] as? String
            ?: throw QuranBridgeException.invalidArgs("reciterId")

        currentSurah = surah
        currentAyah  = map["fromAyah"] as? Int

        emit("loading")
        stop()

        // TODO: integrate quran_android logic here — resolve audio URL from reciter + surah
        val url = resolveAudioUrl(reciter, surah)

        player = MediaPlayer().apply {
            setDataSource(url)
            setOnPreparedListener {
                it.start()
                emit("playing")
                handler.post(progressRunnable)
            }
            setOnCompletionListener {
                handler.removeCallbacks(progressRunnable)
                emit("stopped")
                release()
                player = null
            }
            setOnErrorListener { _, what, extra ->
                handler.removeCallbacks(progressRunnable)
                emitError("MediaPlayer error: what=$what extra=$extra")
                true
            }
            prepareAsync()
        }
    }

    fun pause() {
        player?.takeIf { it.isPlaying }?.pause()
        handler.removeCallbacks(progressRunnable)
        emit("paused")
    }

    fun stop() {
        handler.removeCallbacks(progressRunnable)
        player?.apply {
            if (isPlaying) stop()
            release()
        }
        player = null
        emit("stopped")
    }

    fun release() {
        stop()
        eventSink = null
    }

    // ── Private ───────────────────────────────────────────────────────────────

    private fun emit(
        status: String,
        positionMs: Int? = null,
        durationMs: Int? = null,
    ) {
        val state = mutableMapOf<String, Any>("status" to status)
        currentSurah?.let { state["surahNumber"] = it }
        currentAyah?.let  { state["ayahNumber"]  = it }
        positionMs?.let   { state["positionMs"]  = it }
        durationMs?.let   { state["durationMs"]  = it }
        eventSink?.success(state)
    }

    private fun emitError(message: String) {
        val state = mutableMapOf<String, Any>(
            "status"       to "error",
            "errorMessage" to message,
        )
        currentSurah?.let { state["surahNumber"] = it }
        eventSink?.success(state)
    }

    // TODO: integrate quran_android logic here — replace with real URL resolution
    private fun resolveAudioUrl(reciter: String, surah: Int): String {
        val paddedSurah = surah.toString().padStart(3, '0')
        return "https://cdn.islamic.network/quran/audio-surah/128/$reciter/$paddedSurah.mp3"
    }
}
